cuqi/distribution/_distribution.py

```diff
@@ -125,8 +124,49 @@ def geometry(self,value):
         else:
             raise TypeError("The attribute 'geometry' should be of type 'int' or 'cuqi.geometry.Geometry', or None.")
 
+    def logd(self, *args, **kwargs):
+        """  Evaluate the log density of the distribution.
+
+        The log density is proportional to the log probability density function (logpdf) of the distribution.
+
+        It is possible to pass conditioning variables as arguments to this function in addition to the parameters of the distribution.
+        """
+
+        # Get the (potential) conditioning variables
+        cond_vars = self.get_conditioning_variables()
+
+        # If distribution is conditional, we first condition before evaluating the log density
+        if len(cond_vars) > 0:
+
+            if len(args) > 0:
+                raise ValueError("Distribution.logd: Positional arguments on conditional distributions are not supported in logp evaluation (yet)")
+
+            # Check if all conditioning variables are specified
+            if not all([key in kwargs for key in cond_vars]):
+                raise ValueError(f"Distribution.logd: To evaluate the log density all conditioning variables must be specified. Conditioning variables are: {cond_vars}")
+
+            # Extract exactly the conditioning variables from kwargs
+            cond_kwargs = {key: kwargs[key] for key in cond_vars}
+
+            # Extract any remaining variables from kwargs
+            non_cond_kwargs = {key: kwargs[key] for key in kwargs if key not in cond_vars}
+
+            # Condition the distribution on the conditioning variables
+            new_dist = self(**cond_kwargs)
+
+            # Now evaluate the log density of the fully specified distribution
+            return new_dist.logd(*args, **non_cond_kwargs)
+
+        # Not conditional distribution, simply evaluate log density directly
+        else:
+            return super().logd(*args, **kwargs)
```

Comment:
Good question. Passing arg to logd should be supported almost always in my opinion. But in some cases we it may not be implemented or have ambigious meaning.

For example for conditional distributions evaluating logd requires providing both the parameter for the distribution and any conditional variables. It is unclear if the parameter or conditioning variables should be first, so currently it simply throws an error.

Reply:
got it. Thanks!